File: Top-Interview-Questions/Algorithms/200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 * Runtime: 68 ms, faster than 56.80% of JavaScript online submissions for Number of Islands.
 * Memory Usage: 37.2 MB, less than 100.00% of JavaScript online submissions for Number of Islands.
 * Time Complexity: O(N), where N is the number of elements in a grid
 * Space Complexity: O(N), where N is the number of elements in a grid
 */

const ISLAND                = '1';
const WATER                 = '0';
const VISITED_ISLAND        = '2';
const H_DIRECTION_INDICATOR = [0, 0, 1, -1];
const V_DIRECTION_INDICATOR = [1, -1, 0, 0];

var numIslands = function(grid) {
    if (!grid) {
        return 0;
    }

    let numOfIsland = 0;
    for (let rowIdx = 0; rowIdx < grid.length; rowIdx++) {
        for (let colIdx = 0; colIdx < grid[0].length; colIdx++) {
            if (grid[rowIdx][colIdx] === ISLAND) {
                numOfIsland++;

                traverseIsland(grid, rowIdx, colIdx);
            }
        }
    }

    return numOfIsland;
};

function traverseIsland (grid, rowIdx, colIdx) {
    if (rowIdx < 0 || rowIdx >= grid.length || colIdx < 0 || colIdx >= grid[0].length) {
        return;
    } else if (grid[rowIdx][colIdx] === VISITED_ISLAND || grid[rowIdx][colIdx] === WATER) {
        return;
    }

    grid[rowIdx][colIdx] = VISITED_ISLAND;

    for (let index = 0; index < 4; index++) {
        traverseIsland(grid, rowIdx + V_DIRECTION_INDICATOR[index], colIdx + H_DIRECTION_INDICATOR[index]);
    }
}

File: May-LeetCoding-Challange-2020/Week3/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        
        count = 0
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j)
                    count = count + 1
        return count
    
    def dfs(self, grid, i, j):
        if i<0 or j<0 or i>=len(grid) or j>= len(grid[0]) or grid[i][j] != '1':
            return
        
        grid[i][j] = '0'
        self.dfs(grid, i+1, j)
        self.dfs(grid, i-1, j)
        self.dfs(grid, i, j+1)
        self.dfs(grid, i, j-1)

    # Time Complexity: O(m*n) => O(n^2)
    # Space Complexity: O(m*n), where the map is filled with all lands
    # Run Code Runtime: 44 ms
    # Submit Runtime: 152 ms, 51% faster
    # Submit Memory: 9.40% less